### PR TITLE
Using x11-rs for X11 bindings.

### PR DESCRIPTION
**Pull request by Daggerbot**

### CODE REVIEW

**Comment by tomaka** [comment]
I think you don't need `mem::transmute` to handle incompatibilities. Just adding `as *const _` or `as *mut _` works and is less error-prone.

File: src/x11/window/mod.rs [hunk 1]
```diff
@@ -69,7 +67,7 @@ impl Drop for XWindow {
         unsafe {
             // we don't call MakeCurrent(0, 0) because we are not sure that the context
             // is still the current one
-            ffi::glx::DestroyContext(self.display, self.context);
+            ffi::glx::DestroyContext(mem::transmute(self.display), self.context);
```
Comment (tomaka) [inline src/x11/window/mod.rs]: Here

File: src/x11/window/mod.rs [hunk 2]
```diff
@@ -124,34 +122,34 @@ impl<'a> Iterator for PollEventsIterator<'a> {
...
                     return None;
                 }
             }
-    
-            match xev.type_ {
+
+            match xev.kind() {
                 ffi::KeymapNotify => {
-                    unsafe { ffi::XRefreshKeyboardMapping(&xev) }
+                    unsafe { ffi::XRefreshKeyboardMapping(mem::transmute(&xev)) }
```
Comment (tomaka) [inline src/x11/window/mod.rs]: Here

File: src/x11/window/mod.rs [hunk 3]
```diff
@@ -383,7 +381,7 @@ impl Window {
         // creating the color map
         let cmap = unsafe {
             let cmap = ffi::XCreateColormap(display, root,
-                visual_infos.visual, ffi::AllocNone);
+                mem::transmute(visual_infos.visual), ffi::AllocNone);
```
Comment (tomaka) [inline src/x11/window/mod.rs]: Here

File: src/x11/window/mod.rs [hunk 4]
```diff
@@ -415,7 +413,7 @@ impl Window {
         let window = unsafe {
             let win = ffi::XCreateWindow(display, root, 0, 0, dimensions.0 as libc::c_uint,
                 dimensions.1 as libc::c_uint, 0, visual_infos.depth, ffi::InputOutput,
-                visual_infos.visual, window_attributes,
+                mem::transmute(visual_infos.visual), window_attributes,
```
Comment (tomaka) [inline src/x11/window/mod.rs]: Here

**Comment by tomaka** [comment]
Bindings for OSMesa are missing.
You should compile with `cargo test --features "headless"` to try it out.

File: src/lib.rs [hunk 5]
```diff
@@ -46,6 +46,10 @@ extern crate cocoa;
 extern crate core_foundation;
 #[cfg(target_os = "macos")]
 extern crate glutin_core_graphics as core_graphics;
+#[cfg(target_os = "linux")]
+extern crate x11;
+#[cfg(target_os = "freebsd")]
+extern crate x11;
```
Comment (SSheldon) [inline src/lib.rs]: I believe these can be combined with:
``` rust
#[cfg(any(target_os = "linux", target_os = "freebsd"))]
extern crate x11;
```

**Pull request closed without merging**